src/command_modules/azure-cli-network/azure/cli/command_modules/network/custom.py

```diff
@@ -92,17 +91,99 @@ def list_application_gateways(resource_group_name=None):
+    virtual_network_name = virtual_network_name or '{}Vnet'.format(application_gateway_name)
+
+    # Build up the ARM template
+    master_template = ArmTemplateBuilder()
+    ag_dependencies = []
+
+    public_ip_id = public_ip_address if is_valid_resource_id(public_ip_address) else None
+    subnet_id = subnet if is_valid_resource_id(subnet) else None
+    private_ip_allocation = IPAllocationMethod.static.value if private_ip_address \
+        else IPAllocationMethod.dynamic.value
+
+    network_id_template = resource_id(
+        subscription=get_subscription_id(), resource_group=resource_group_name,
+        namespace='Microsoft.Network')
+
+    if subnet_type == 'new':
+        ag_dependencies.append('Microsoft.Network/virtualNetworks/{}'.format(virtual_network_name))
+        vnet = build_vnet_resource(
+            virtual_network_name, location, tags, vnet_address_prefix, subnet,
+            subnet_address_prefix)
+        master_template.add_resource(vnet)
+        subnet_id = '{}/virtualNetworks/{}/subnets/{}'.format(network_id_template,
+                                                              virtual_network_name, subnet)
+
+    if public_ip_address_type == 'new':
+        ag_dependencies.append('Microsoft.Network/publicIpAddresses/{}'.format(public_ip_address))  # pylint: disable=line-too-long
+        master_template.add_resource(build_public_ip_resource(public_ip_address, location,
+                                                              tags,
+                                                              public_ip_address_allocation,
+                                                              None))
+        public_ip_id = '{}/publicIPAddresses/{}'.format(network_id_template,
+                                                        public_ip_address)
+
+    app_gateway_resource = build_application_gateway_resource(
+        application_gateway_name, location, tags, sku, sku_tier, capacity, servers, frontend_port,
+        private_ip_address, private_ip_allocation, cert_data, cert_password,
+        http_settings_cookie_based_affinity, http_settings_protocol, http_settings_port,
+        http_listener_protocol, routing_rule_type, public_ip_id, subnet_id)
+    app_gateway_resource['dependsOn'] = ag_dependencies
+    master_template.add_variable(
+        'appGwID',
+        "[resourceId('Microsoft.Network/applicationGateways', '{}')]".format(
+            application_gateway_name))
+    master_template.add_resource(app_gateway_resource)
+    master_template.add_output('applicationGateway', application_gateway_name, output_type='object')
+
+    template = master_template.build()
+
+    # deploy ARM template
+    deployment_name = 'ag_deploy_' + random_string(32)
+    client = get_mgmt_service_client(ResourceManagementClient).deployments
+    properties = DeploymentProperties(template=template, parameters={}, mode='incremental')
+    if validate:
+        from pprint import pprint
+        pprint(template)
```

Comment:
you can consider to reuse https://github.com/Azure/azure-cli/blob/master/src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/_vm_utils.py#L49